src/journal/views.py

```diff
@@ -1392,36 +1392,73 @@ def search(request):
 
-    if request.POST:
-        search_term = request.POST.get('search')
-        request.session['article_search'] = search_term
-        return redirect(reverse('search'))
+    search_term, keyword, sort, redir = logic.handle_search_controls(request)
 
-    if request.session.get('article_search'):
-        search_term = request.session.get('article_search')
+    if redir:
+        return redir
+    #if no order, set order to title (alphabetical). this happens when receiveing a GET request for keyword.
+    if request.GET and not sort:
+        sort='title'
 
+    if search_term:
+        # return two lists of pks based on search term.
         article_search = submission_models.Article.objects.filter(
             (Q(title__icontains=search_term) |
-             Q(subtitle__icontains=search_term)) &
-            Q(journal=request.journal)
+             Q(keywords__word__icontains=search_term) |
+             Q(subtitle__icontains=search_term)), 
+            journal=request.journal, 
+            stage=submission_models.STAGE_PUBLISHED,
+            date_published__lte=timezone.now()
         )
-        article_search = [article for article in article_search]
+        article_search_unsorted = [article.pk for article in article_search]
 
-        author_search = search_term.split(' ')
         from_author = submission_models.FrozenAuthor.objects.filter(
-            (Q(first_name__in=author_search) |
-             Q(last_name__in=author_search)) &
-            Q(article__journal=request.journal)
+            (Q(first_name__icontains=search_term) |
+             Q(last_name__icontains=search_term)),
+            article__journal=request.journal, 
+            article__stage=submission_models.STAGE_PUBLISHED, 
+            article__date_published__lte=timezone.now()
         )
-
-        articles_from_author = [author.article for author in from_author]
-        articles = set(article_search + articles_from_author)
+        articles_from_author_unsorted = [author.article.pk for author in from_author]
+        # merge the two lists into a set which will clear out duplicates
+        article_list_unsorted = set(article_search_unsorted + articles_from_author_unsorted)
+        # do a search on the set of pks and apply the order_by while turning back into a list of article objects.
+        articles = list(submission_models.Article.objects.filter(pk__in=article_list_unsorted).order_by(sort))
+
+    # just single keyword atm. but keyword is included in article_search.
+    elif keyword:
+        keyword_search = submission_models.Article.objects.filter(
+            keywords__word=keyword, 
+            journal=request.journal, 
+            stage=submission_models.STAGE_PUBLISHED, 
+            date_published__lte=timezone.now()
+        ).order_by(sort)
+        articles = [article for article in keyword_search]
+
+    # all keywords of published articles.
+    published_articles = submission_models.Article.objects.filter(
```

Comment:
Unless I'm missing something, there is no need for the double database roundtrip here, you can apply the article filters onto `Keyword.objects.filter()` call below.

Reply:
if there is a `search_term` performs a search based on title, keyword, subtitle, and author.
otherwise if there is a `keyword` it just searches for all articles that have that keyword. This was set up so you can kind of have an ontology across the site and organize things by theme.

So you have a Journal that deals with Cities and have an article titled "The big green lizard who wants to destroy Tokyo" with keywords of "Japan", "Urban Planning", "Disaster", etc. Japan and Ubran planning aren't available in the title or maybe even the subtitle but they would get picked up here. And if those keywords are available on the article then you would be able to click those links and see these results.

The `Keyword.objects.filter()` call below gets a list of keywords for the search page and returns those as an option in case you don't know what to search for you can browse these keywords. Your next comment does ask for these to be paired down and that is very legitimate. I'll comment my thoughts on that and how it relates to this below.

Reply:
Yes, I agree with you here, in this comment I was only referring to the queries for `published_articles` and `all_keywords` which could be merged, (unless you are trying to reduce the cardinality)